receiver/awss3rehydrationreceiver/reciever.go

```diff
@@ -288,31 +339,35 @@ func (r *rehydrationReceiver) processObject(object *aws.ObjectInfo) error {
 		return fmt.Errorf("unsupported file type: %s", ext)
 	}
 
-	if err := r.consumer.Consume(r.ctx, objectBuffer); err != nil {
+	if err := r.consumer.Consume(ctx, objectBuffer); err != nil {
 		return fmt.Errorf("consume: %w", err)
 	}
 
 	return nil
 }
 
-// checkpointStorageKey is the key used for storing the checkpoint
-const checkpointStorageKey = "aws_s3_checkpoint"
+func (r *rehydrationReceiver) saveCheckpointData(objectTime *time.Time, objectName string) {
+	r.checkpointMutex.Lock()
+	defer r.checkpointMutex.Unlock()
 
-// checkpointKey returns the key used for storing the checkpoint
-func (r *rehydrationReceiver) checkpointKey() string {
-	return fmt.Sprintf("%s_%s_%s", checkpointStorageKey, r.id, r.supportedTelemetry.String())
+	if r.lastObjectTime == nil || r.lastObjectTime.Before(*objectTime) {
+		r.lastObjectName = objectName
+		r.lastObjectTime = objectTime
+	}
 }
 
-// checkEntityCount checks the number of entities rehydrated and the current state of the
-// empty counter. If zero entities were rehydrated increment the counter.
-// If there were entities rehydrated reset the counter as we want to track consecutive zero sized polls.
-func checkEntityCount(numEntitiesRehydrated, emptyEntityCounter int) int {
-	switch {
-	case emptyEntityCounter == emptyPollLimit: // If we are at the limit return the limit
-		return emptyPollLimit
-	case numEntitiesRehydrated == 0: // If no entities were rehydrated then increment the empty entities counter
-		return emptyEntityCounter + 1
-	default: // Default case is numEntitiesRehydrated > 0 so reset emptyEntityCounter to 0
-		return 0
+func (r *rehydrationReceiver) handleCheckpoint(ctx context.Context) error {
+	r.checkpointMutex.Lock()
+	defer r.checkpointMutex.Unlock()
+
+	if r.lastObjectName == "" || r.lastObjectTime == nil {
+		return nil
 	}
+
+	// update && store checkpoint
+	r.checkpoint.UpdateCheckpoint(*r.lastObjectTime, r.lastObjectName)
+	if err := r.checkpointStore.SaveCheckpoint(ctx, r.checkpointKey, r.checkpoint); err != nil {
+		return fmt.Errorf("save checkpoint: %w", err)
+	}
+	return nil
```

Comment:
```suggestion
	if r.lastObjectName == "" || r.lastObjectTime == nil {
		return nil
	}
	r.checkpointMutex.Lock()
	defer r.checkpointMutex.Unlock()

	// update && store checkpoint
	r.checkpoint.UpdateCheckpoint(*r.lastObjectTime, r.lastObjectName)
	if err := r.checkpointStore.SaveCheckpoint(ctx, r.checkpointKey, r.checkpoint); err != nil {
		return fmt.Errorf("save checkpoint: %w", err)
	}
	return nil
```

Nit: We probably don't want to lock the mutex if we are doing an early return in this case.

Reply:
Yea that's fair, there's no consequential race condition if we end up returning early